cibuildwheel/windows.py

```diff
@@ -313,22 +313,32 @@ def build(options: BuildOptions) -> None:
                 config_setting = " ".join(verbosity_flags)
                 build_env = env.copy()
                 if options.dependency_constraints:
-                    build_env["PIP_CONSTRAINT"] = str(
-                        options.dependency_constraints.get_for_python_version(config.version)
+                    constr = options.dependency_constraints.get_for_python_version(config.version)
+                    # Bug in pip <= 21.1.3 - we can't have a space in the
+                    # constraints file, and pip doesn't support drive letters
+                    # in uhi.  After probably pip 21.2, we can use uri. For
+                    # now, use a temporary file.
+                    if " " in str(constr):
+                        with tempfile.NamedTemporaryFile(
+                            "w", suffix="constraints.txt", delete=False
```

Comment:
Why not setting `dir="C:\\cibw\\"` ?
We're already installing python in there and we know for sure there's no space in that path.

Reply:
Should the file be deleted ? or we just don't care to clean up running in CI and all.

Reply:
Well, eventually I'd like to not install to absolute paths, to make it nicer to run on local machines, and requiring install in drive letter C would also also be nice to avoid. But we could for now. I _think_ default temporary directories avoid spaces in their names?

Reply:
It's important to close the file before reading from it on Windows, which is why I have this here -  don't think I can close but not delete without it. I was trying to tell if this gets deleted anyway, but that wasn't clear from some perusal of tmpfile.py. I wasn't that worried about leaving it around, but it would be a little nicer to try to delete it if it's not deleted already.

Reply:
> I think default temporary directories avoid spaces in their names?

I wouldn't bet on that on windows.

If the test passes CI, that means it's true on those images (and it will probably remain that way for CI images, space in path is often a problem). For running on local machines, I'm not so sure it won't happen.

> Well, eventually I'd like to not install to absolute paths, to make it nicer to run on local machines, and requiring install in drive letter C would also also be nice to avoid.

Agreed, also, seeing this comment https://github.com/pypa/packaging.python.org/pull/897#discussion_r664532138, I thought we might be close to being able to run "stateless" on windows however, in this specific instance, I don't think it's a bad idea to use a hardcoded path as long as it's the same as the python install dir, maybe using a top-level "constant" to be able to get this consistently for now (and at least, it won't be forgotten if we go further the "stateless" road).

Reply:
> Why not use keyword arg dir="C:\\cibw\\" ?
> We're already installing python in there and we know for sure there's no space in that path.

Sounds like a good idea to me. Looking at [the implementation of `tempfile`](https://github.com/python/cpython/blob/9761abf30617946a374886184b0e36ec050dd8d7/Lib/tempfile.py#L158), there are plenty of places that a space might appear in that code. 

> Should the file be deleted ?

I think it's okay to leave it lying around, as it makes the implementation easier, and this should hopefully be a temporary fix.

Reply:
Honestly, since this installs to `C:/cibw`, it's not that bad to run it locally. It's not affecting anything but itself. It would be nicer to install somewhere more local and nicer to not require a C drive, but it's not replacing system Pythons or anything. It's really just macOS that is a very bad idea to run locally.

I've pulled this into a top level constant, and set the dir here.